### DISCUSSION TITLE
Is it possible to format money to europe in the form builder?

### DISCUSSION THREAD

**Discussion by nielsvh2103** [filamentphp/filament#12892] [Help] [answered]
### Package

Form builder

### Package Version

v3.2.79

### How can we help you?

Hi there,

We live in europe (the netherlands) and we are possibly having an issue with the form builder.

In tables you are able to do the following:

```                
Tables\Columns\TextColumn::make('price')
                    ->label('Prijs')
                    ->money('EUR', locale: 'nl'), 
```

Now in forms all you have is inputMode('decimal'), but that does not suffice for a price field.

In the docs and the only way i found to force 2 decimals is by using the following:

```
                  TextInput::make('price')
                                    ->prefix('€')
                                    ->mask(RawJs::make('$money($input)'))
                                    ->stripCharacters(',')
                                    ->numeric()
                                    ->nullable(),
```

The problem with this is, as we do live in europe, you can not set this to your app locale or give a specific format.

The moment you start typing 2.000 becomes 2,000
We have to reverse this, but we found no way to do this in the form builder currently.

I was wondering if there is anything to do this, or in that case, add this to filament?
methods that would be helpful for example;

decimalSeparator()
thousandSeparator()
toFixed(2) , or for exapmple toFixed(3) (for 3 decimals)

What would work for us in this case is also just ->money('EUR') on this input so it reverses comma and dots.

Either way we are wondering if it is 'currently' possible to do this already, if not if it could be added as money inputs are pretty common.

Thanks in advance :)

**Comment by D-Jordy** [comment]
you can cast the value (https://filamentphp.com/docs/3.x/panels/getting-started#casting-the-price-to-an-integer)

this will also be reflected in the input, add in a `->prefix('€')` and done

**Reply by nielsvh2103** [reply]
Hi there,

I know you can do this, but we have specific needs to limit the user of filling specifics.

As also with `RawJs::make('$money($input)')`
this makes sure the thousandseparator is automatically added when typing and limits you to 2 decimals.

The problem here is i can not format it to EUR

Just casting the value won't set the specific requirements i need for a money input

My examples:
€100,20 is possible , 100,20000 is not possible. - by just casting , you can still type this value (we want to prevent this)
€1.000,20 is possible aswell, but typing it should automatically add separators what `RawJs::make('$money($input)')` this piece of code also does. Rather, i can not add a format and the problem is it displays as follows:

![Screenshot 2024-05-22 at 11 24 43](https://github.com/filamentphp/filament/assets/60341437/16352bc6-3b78-455e-aa2b-220447357956)

`RawJs::make('$money($input)')`  This definitely works, though its in US format, not EU.

**Reply by D-Jordy** [reply]
that is something you will have resolve on the JS side.
 you can use the Alpine mask and use coma's instead of periods (https://alpinejs.dev/plugins/mask#money-inputs) `$money($input, ',')` should do

**Reply by nielsvh2103** [reply]
Thanks i was hoping something like that would work and i did google for alpine magic methods
Seems it is in there, though i just did the following:

` ->mask(RawJs::make('$money($input, ",")'))`

Instead of

` ->mask(RawJs::make('$money($input)'))`

Though, now it completely doesn't work anymore. (if i add a 2nd argument)

Full input:
```
                    TextInput::make('price')
                                    ->prefix('€')
                                    ->inputMode('decimal')
                                    ->mask(RawJs::make('$money($input, ",")'))
                                    ->numeric()
                                    ->nullable(),
```

**Reply by nielsvh2103** [reply]
Coming back to myself i just debugged in the vendor map

Going to input.blade.php 

    x-mask:dynamic="$money($input, ',')"
Adding this myself = it works

The merge seems to go wrong it seems, because this:
Doesnt work while it doe slook like it should;
`    'x-mask' . ($mask instanceof \Filament\Support\RawJs ? ':dynamic' : '') => filled($mask) ? $mask : null,`

```
        <x-filament::input
            :attributes="
                \Filament\Support\prepare_inherited_attributes($getExtraInputAttributeBag())
                    ->merge($extraAlpineAttributes, escape: false)
                    ->merge([
                        'autocapitalize' => $getAutocapitalize(),
                        'autocomplete' => $getAutocomplete(),
                        'autofocus' => $isAutofocused(),
                        'disabled' => $isDisabled,
                        'id' => $id,
                        'inlinePrefix' => $isPrefixInline && (count($prefixActions) || $prefixIcon || filled($prefixLabel)),
                        'inlineSuffix' => $isSuffixInline && (count($suffixActions) || $suffixIcon || filled($suffixLabel)),
                        'inputmode' => $getInputMode(),
                        'list' => $datalistOptions ? $id . '-list' : null,
                        'max' => (! $isConcealed) ? $getMaxValue() : null,
                        'maxlength' => (! $isConcealed) ? $getMaxLength() : null,
                        'min' => (! $isConcealed) ? $getMinValue() : null,
                        'minlength' => (! $isConcealed) ? $getMinLength() : null,
                        'placeholder' => $getPlaceholder(),
                        'readonly' => $isReadOnly(),
                        'required' => $isRequired() && (! $isConcealed),
                        'step' => $getStep(),
                        'type' => $type,
                        $applyStateBindingModifiers('wire:model') => $statePath,
                        'x-bind:type' => $isPasswordRevealable ? 'isPasswordRevealed ? \'text\' : \'password\'' : null,
                        'x-mask' . ($mask instanceof \Filament\Support\RawJs ? ':dynamic' : '') => filled($mask) ? $mask : null,
                    ], escape: false)
                    ->class([
                        '[&::-ms-reveal]:hidden' => $isPasswordRevealable,
                    ])
            "
        />
```

Seems to be a bug?

**Reply by D-Jordy** [reply, marked as answer]
could try adding this to the input: `->extraAlpineAttributes([
            'x-mask:dynamic' => '$money($input, \',\')'])`
            
unsure about your case beeing a bug

**Reply by nielsvh2103** [reply]
Yeah this works, though weird the mask is not properly being set in my case.

It is inside a repeater though.

Edit:

Its fixed and after debugging it was not a bug after all. 
`->mask(RawJs::make('$money($input, \',\')'))`
Is the correct way and seems to work for me now :)

Earlier i showed;

`      ->mask(RawJs::make('$money($input, ",")'))`
This doesn't work as it does not automatically escape properly for you, so you have to use this
`$money($input, \',\')` 

Thanks!
